Port WebExtension Commands to C++

ItsJamie9494 opened the pull request:
#### 3bc3dc091ab0120f61753f3ea5ceb29968dbab56
<pre>
Port WebExtension Commands to C++
<a href="https://webkit.org/b/282876">https://webkit.org/b/282876</a>

Reviewed by Timothy Hatcher.

Port WebExtension commands to C++, and add in support for Linux-specific suggested shortcuts.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::commands): Deleted.
(WebKit::WebExtension::hasCommands): Deleted.
(WebKit::parseCommandShortcut): Deleted.
(WebKit::WebExtension::populateCommandsIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::commands):
(WebKit::WebExtension::hasCommands):
(WebKit::parseCommandShortcut):
(WebKit::WebExtension::populateCommandsIfNeeded):

Canonical link: <a href="https://commits.webkit.org/286420@main">https://commits.webkit.org/286420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8d17b2010783896a8d78ec8cdbc33c0b924dbc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80456 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27225 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64130 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3284 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59559 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17717 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79026 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39919 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46839 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25552 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67954 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23051 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81920 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3328 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2113 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3482 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67097 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11043 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9164 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11745 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3278 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6084 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3299 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4237 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3306 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1900,6 +1906,227 @@ RefPtr<WebCore::Icon> WebExtension::bestIconVariantForManifestKey(const JSON::Ob
+
+                String platformShortcut;
+#if PLATFORM(MAC)
+                platformShortcut = !suggestedKeyObject->getString(macPlatform).isEmpty() ? suggestedKeyObject->getString(macPlatform) : suggestedKeyObject->getString(iosPlatform);
+#elif PLATFORM(IOS_FAMILY)
+                platformShortcut = !suggestedKeyObject->getString(iosPlatform).isEmpty() ? suggestedKeyObject->getString(iosPlatform) : suggestedKeyObject->getString(macPlatform);
```
I think we want to simply null check the string. That way a platform can be defined as empty with no fallback.

```suggestion
                platformShortcut = suggestedKeyObject->getString(macPlatform) ?: suggestedKeyObject->getString(iosPlatform);
#elif PLATFORM(IOS_FAMILY)
                platformShortcut = suggestedKeyObject->getString(iosPlatform) ?: suggestedKeyObject->getString(macPlatform);
```

ItsJamie9494 replied:
as far as I can tell WTF::String doesn't have a default boolean operator, so we can't simply null-check the strings, we have to check isNull or isEmpty

xeenon replied:
Oh right, only explicit bool operator works with String.

ItsJamie9494 committed:
Port WebExtension Commands to C++
https://webkit.org/b/282876

Reviewed by Timothy Hatcher.

Port WebExtension commands to C++, and add in support for Linux-specific suggested shortcuts.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::commands): Deleted.
(WebKit::WebExtension::hasCommands): Deleted.
(WebKit::parseCommandShortcut): Deleted.
(WebKit::WebExtension::populateCommandsIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::commands):
(WebKit::WebExtension::hasCommands):
(WebKit::parseCommandShortcut):
(WebKit::WebExtension::populateCommandsIfNeeded):

Canonical link: https://commits.webkit.org/286420@main
